pkg/clusteragent/admission/mutate/auto_instrumentation.go

```diff
@@ -194,24 +206,87 @@ func injectApmTelemetryConfig(pod *corev1.Pod) {
 	_ = injectEnv(pod, instrumentationInstallIDEnvVar)
 }
 
-func getAllLibsToInject(registry string) map[string]libInfo {
-	libsToInject := map[string]libInfo{}
+// getLibrariesToInjectForApmInstrumentation returns the list of tracing libraries to inject, when APM Instrumentation is enabled
+// - if apm_config.instrumentation.lib_versions set, returns only tracing libraries from apm_config.instrumentation.lib_versions
+// - if language detection is on and can detect the apps' languages, returns only auto-detected languages
+// - otherwise returns all tracing libraries supported by APM Instrumentation
+func getLibrariesToInjectForApmInstrumentation(pod *corev1.Pod, registry string) ([]libInfo, bool) {
+	autoDetected := false
+
+	// Pinned tracing libraries in APM Instrumentation configuration
+	libsToInject := getPinnedLibraries(registry)
+	if len(libsToInject) > 0 {
+		return libsToInject, autoDetected
+	}
+
+	// Tracing libraries for language detection
+	libsToInject = getLibrariesLanguageDetection(pod, registry)
+	if len(libsToInject) > 0 {
+		autoDetected = true
+		return libsToInject, autoDetected
+	}
+
+	// Latest tracing libraries for all supported languages (java, js, dotnet, python, ruby)
+	libsToInject = getAllLatestLibraries(registry)
+
+	return libsToInject, autoDetected
+}
+
+// getPinnedLibraries returns tracing libraries to inject as configured by apm_config.instrumentation.lib_versions
+func getPinnedLibraries(registry string) []libInfo {
+	// check if pinned libraries have been cached
+	pinnedLibs.libsMutex.Lock()
+	libsToInject := pinnedLibs.libraries
+	pinnedLibs.libsMutex.Unlock()
```

Comment:
To ensure thread safety if at some point of time webhook logic will run in multiple threads. Might be an overkill since this method will be run during Admission Controller creation after @davidor's refactoring, but until then it provides safety against introducing new threads that can modify list of pinned libraries.

Reply:
Reads from map are thread safe. We need to only protect writes.

Reply:
[sync.Map](https://pkg.go.dev/sync#Map) would be safe for concurrent access. Here we are using a slice and assignment is not thread safe; so we can’t safely read from slice X if we’re appending to it via X = append(X, …).